### PR TITLE
[FLINK-8473][webUI] Improve error behavior of JarListHandler

### PR DESCRIPTION
**Pull request by zentol**
## What is the purpose of the change

This PR modifies the `JarListHandler` to not crash when the jar storage directory has been deleted. This caused the job-submission-page in the webUI to be completely blank.

We now log a warning (to make debugging easier) and continue on as if the directory was empty (so that the page isn't blank).

## Verifying this change

Manually verified and not covered by tests, like the rest of the handler.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarListHandler.java [hunk 1]
```diff
@@ -145,6 +155,7 @@ public boolean accept(File dir, String name) {
 					return writer.toString();
 				}
 				catch (Exception e) {
+					LOG.warn("Failed to fetch jar list.", e);
```
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarListHandler.java]: Aren't failed completions logged anyways?

Comment (zentol) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarListHandler.java]: not in legacy handlers

Comment (zentol) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarListHandler.java]: scratch that, should be logged, I'll double check

File: flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarListHandler.java [hunk 2]
```diff
@@ -77,6 +82,11 @@ public boolean accept(File dir, String name) {
 						}
 					});
 
+					if (list == null) {
+						LOG.warn("Jar storage directory {} has been deleted.", jarDir.getAbsolutePath());
```
Comment (uce) [inline flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarListHandler.java]: Somehow this comment got lost before: Should we make the error message more explicit and say that it was `deleted externally` (e.g. not by Flink)?

Comment (zentol) [reply flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/JarListHandler.java]: yes thats a good idea

**Comment by zentol** [comment]
@uce Could you take another look?

**Comment by uce** [comment]
👍  to merge.

**Comment by StephanEwen** [comment]
Will this allow new upload of jars? Would we need an `mkdirs` to re-create that directory?

**Comment by zentol** [comment]
I haven't tried that out yet. I couldn't quite figure out where the upload happens (maybe completely handled by netty?).

**Comment by uce** [comment]
I just tried it, re-uploading after deleting the directory **does not work**. Good catch Stephan. :-)

@zentol: I found `HttpRequestHandler` which handles the uploads. The handler assumes that the directory exists and the logic for creating the temp directory is in `WebRuntimeMonitor`. We should consolidate this in a single place (if no one else needs this directory).

**Comment by zentol** [comment]
I've updated the PR. All handlers that access the jar storage directory now check whether it exists and regenerate it if necessary with appropriate logging messages. The logging and directory creation is done via static methods in the WebRuntimeMonitor.

**Comment by StephanEwen** [comment]
Thanks, looks good!

Please merge into `master` and `release-1.4`...

**Comment by zentol** [comment]
merging.

**Pull request closed without merging**